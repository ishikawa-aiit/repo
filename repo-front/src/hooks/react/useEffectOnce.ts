import { useEffect, useRef, useState } from 'react'

/**
 * 依存関係がゼロの時、useEffect（）の代わりに使用するカスタムフック
 *
 * @remarks
 * このカスタムフックはReact 18の仕様変更のため、作成しています（React 18では依存関係がゼロの時、2回レンダリングされる）
 * https://github.com/vercel/next.js/issues/35822
 */
export const useEffectOnce = (effect: () => void | (() => void)) => {
  const destroyFunc = useRef<void | (() => void)>()
  const effectCalled = useRef(false)
  const renderAfterCalled = useRef(false)
  const [_val, setVal] = useState<number>(0)

  if (effectCalled.current) {
    renderAfterCalled.current = true
  }

  useEffect(() => {
    if (!effectCalled.current) {
      destroyFunc.current = effect()
      effectCalled.current = true
    }

    setVal((val) => {
      return val + 1
    })

    return () => {
      if (!renderAfterCalled.current) {
        return
      }
      if (destroyFunc.current) {
        destroyFunc.current()
      }
    }
    // マウント時に1回だけ実行するため
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])
}
