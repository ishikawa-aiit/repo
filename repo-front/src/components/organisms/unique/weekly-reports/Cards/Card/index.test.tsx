import { render, screen, within } from '@testing-library/react'

import { Card } from '.'

const testData = {
  reportWeek: '07/04 ~ 07/09',
  activityTime: 'テスト活動時間',
  doneActivity: 'やったこと',
  todoActivity: 'やること',
  solution: '解決したこと',
  event: '起こったこと',
  remark: '特別なこと',
}

describe('Card', () => {
  it('正常にレンダリングされている', () => {
    const { container } = render(<Card weeklyReportItem={testData} id={1} />)
    const { getAllByRole } = within(screen.getByTestId('Card'))

    expect(getAllByRole('term')[0]).toHaveTextContent('報告週')
    expect(getAllByRole('term')[1]).toHaveTextContent('活動時間')
    expect(getAllByRole('term')[2]).toHaveTextContent('今週の活動と成果の実績')
    expect(getAllByRole('term')[3]).toHaveTextContent('来週の活動と成果の予定')
    expect(getAllByRole('term')[4]).toHaveTextContent('課題と解決策')
    expect(getAllByRole('term')[5]).toHaveTextContent('できごと・気づき')
    expect(getAllByRole('term')[6]).toHaveTextContent('特記事項')
    expect(getAllByRole('definition')[0]).toHaveTextContent(testData.reportWeek)
    expect(getAllByRole('definition')[1]).toHaveTextContent(testData.activityTime)
    expect(getAllByRole('definition')[2]).toHaveTextContent(testData.doneActivity)
    expect(getAllByRole('definition')[3]).toHaveTextContent(testData.todoActivity)
    expect(getAllByRole('definition')[4]).toHaveTextContent(testData.solution)
    expect(getAllByRole('definition')[5]).toHaveTextContent(testData.event)
    expect(getAllByRole('definition')[6]).toHaveTextContent(testData.remark)
    expect(container).toMatchSnapshot()
  })
})
