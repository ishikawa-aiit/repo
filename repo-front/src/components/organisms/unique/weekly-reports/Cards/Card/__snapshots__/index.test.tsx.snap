// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card 正常にレンダリングされている 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-padding-start: 6px;
  padding-inline-start: 6px;
  -webkit-padding-end: 6px;
  padding-inline-end: 6px;
  padding-top: 3px;
  padding-bottom: 3px;
  border: 1px;
  border-color: neutral.200;
  border-radius: 12px;
  width: 100%;
  background-color: bg.white;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
}

.emotion-1>*:not(style)~*:not(style) {
  margin-top: 0.5rem;
  -webkit-margin-end: 0px;
  margin-inline-end: 0px;
  margin-bottom: 0px;
  -webkit-margin-start: 0px;
  margin-inline-start: 0px;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  border-bottom: 1px;
  border-color: neutral.200;
  width: 100%;
  word-break: break-all;
  padding-top: 2px;
  padding-bottom: 4px;
  gap: 4px;
}

.emotion-2:last-of-type {
  border-bottom: none;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  font-weight: bold;
}

.emotion-4 {
  vertical-align: middle;
  white-space: pre-wrap;
}

.emotion-7 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  position: relative;
  white-space: nowrap;
  vertical-align: middle;
  outline: 2px solid transparent;
  outline-offset: 2px;
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.emotion-8 {
  width: 5px;
}

<div>
  <div
    class="emotion-0"
    data-cy="Card"
    data-testid="Card"
    id="1"
    role="listitem"
  >
    <div
      class="chakra-stack emotion-1"
    >
      <dl
        class="emotion-2"
        data-cy="CardItem"
        data-testid="CardItem"
      >
        <dt
          aria-label="title"
          class="emotion-3"
        >
          報告週
        </dt>
        <dd
          class="chakra-text emotion-4"
        >
          07/04 ~ 07/09
        </dd>
      </dl>
      <dl
        class="emotion-2"
        data-cy="CardItem"
        data-testid="CardItem"
      >
        <dt
          aria-label="title"
          class="emotion-3"
        >
          活動時間
          <button
            class="chakra-button emotion-7"
            data-cy="ClipboardCopyButton"
            data-testid="ClipboardCopyButton"
            type="button"
          >
            <span
              class="chakra-text emotion-8"
            >
              <svg
                aria-hidden="true"
                fill="none"
                stroke="currentColor"
                stroke-width="2"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8 5H6a2 2 0 00-2 2v12a2 2 0 002 2h10a2 2 0 002-2v-1M8 5a2 2 0 002 2h2a2 2 0 002-2M8 5a2 2 0 012-2h2a2 2 0 012 2m0 0h2a2 2 0 012 2v3m2 4H10m0 0l3-3m-3 3l3 3"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
            </span>
          </button>
        </dt>
        <dd
          class="chakra-text emotion-4"
        >
          テスト活動時間
        </dd>
      </dl>
      <dl
        class="emotion-2"
        data-cy="CardItem"
        data-testid="CardItem"
      >
        <dt
          aria-label="title"
          class="emotion-3"
        >
          今週の活動と成果の実績
          <button
            class="chakra-button emotion-7"
            data-cy="ClipboardCopyButton"
            data-testid="ClipboardCopyButton"
            type="button"
          >
            <span
              class="chakra-text emotion-8"
            >
              <svg
                aria-hidden="true"
                fill="none"
                stroke="currentColor"
                stroke-width="2"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8 5H6a2 2 0 00-2 2v12a2 2 0 002 2h10a2 2 0 002-2v-1M8 5a2 2 0 002 2h2a2 2 0 002-2M8 5a2 2 0 012-2h2a2 2 0 012 2m0 0h2a2 2 0 012 2v3m2 4H10m0 0l3-3m-3 3l3 3"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
            </span>
          </button>
        </dt>
        <dd
          class="chakra-text emotion-4"
        >
          やったこと
        </dd>
      </dl>
      <dl
        class="emotion-2"
        data-cy="CardItem"
        data-testid="CardItem"
      >
        <dt
          aria-label="title"
          class="emotion-3"
        >
          来週の活動と成果の予定
          <button
            class="chakra-button emotion-7"
            data-cy="ClipboardCopyButton"
            data-testid="ClipboardCopyButton"
            type="button"
          >
            <span
              class="chakra-text emotion-8"
            >
              <svg
                aria-hidden="true"
                fill="none"
                stroke="currentColor"
                stroke-width="2"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8 5H6a2 2 0 00-2 2v12a2 2 0 002 2h10a2 2 0 002-2v-1M8 5a2 2 0 002 2h2a2 2 0 002-2M8 5a2 2 0 012-2h2a2 2 0 012 2m0 0h2a2 2 0 012 2v3m2 4H10m0 0l3-3m-3 3l3 3"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
            </span>
          </button>
        </dt>
        <dd
          class="chakra-text emotion-4"
        >
          やること
        </dd>
      </dl>
      <dl
        class="emotion-2"
        data-cy="CardItem"
        data-testid="CardItem"
      >
        <dt
          aria-label="title"
          class="emotion-3"
        >
          課題と解決策
          <button
            class="chakra-button emotion-7"
            data-cy="ClipboardCopyButton"
            data-testid="ClipboardCopyButton"
            type="button"
          >
            <span
              class="chakra-text emotion-8"
            >
              <svg
                aria-hidden="true"
                fill="none"
                stroke="currentColor"
                stroke-width="2"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8 5H6a2 2 0 00-2 2v12a2 2 0 002 2h10a2 2 0 002-2v-1M8 5a2 2 0 002 2h2a2 2 0 002-2M8 5a2 2 0 012-2h2a2 2 0 012 2m0 0h2a2 2 0 012 2v3m2 4H10m0 0l3-3m-3 3l3 3"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
            </span>
          </button>
        </dt>
        <dd
          class="chakra-text emotion-4"
        >
          解決したこと
        </dd>
      </dl>
      <dl
        class="emotion-2"
        data-cy="CardItem"
        data-testid="CardItem"
      >
        <dt
          aria-label="title"
          class="emotion-3"
        >
          できごと・気づき
          <button
            class="chakra-button emotion-7"
            data-cy="ClipboardCopyButton"
            data-testid="ClipboardCopyButton"
            type="button"
          >
            <span
              class="chakra-text emotion-8"
            >
              <svg
                aria-hidden="true"
                fill="none"
                stroke="currentColor"
                stroke-width="2"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8 5H6a2 2 0 00-2 2v12a2 2 0 002 2h10a2 2 0 002-2v-1M8 5a2 2 0 002 2h2a2 2 0 002-2M8 5a2 2 0 012-2h2a2 2 0 012 2m0 0h2a2 2 0 012 2v3m2 4H10m0 0l3-3m-3 3l3 3"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
            </span>
          </button>
        </dt>
        <dd
          class="chakra-text emotion-4"
        >
          起こったこと
        </dd>
      </dl>
      <dl
        class="emotion-2"
        data-cy="CardItem"
        data-testid="CardItem"
      >
        <dt
          aria-label="title"
          class="emotion-3"
        >
          特記事項
          <button
            class="chakra-button emotion-7"
            data-cy="ClipboardCopyButton"
            data-testid="ClipboardCopyButton"
            type="button"
          >
            <span
              class="chakra-text emotion-8"
            >
              <svg
                aria-hidden="true"
                fill="none"
                stroke="currentColor"
                stroke-width="2"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8 5H6a2 2 0 00-2 2v12a2 2 0 002 2h10a2 2 0 002-2v-1M8 5a2 2 0 002 2h2a2 2 0 002-2M8 5a2 2 0 012-2h2a2 2 0 012 2m0 0h2a2 2 0 012 2v3m2 4H10m0 0l3-3m-3 3l3 3"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
            </span>
          </button>
        </dt>
        <dd
          class="chakra-text emotion-4"
        >
          特別なこと
        </dd>
      </dl>
    </div>
  </div>
</div>
`;
